### ISSUE TITLE
Removing mmap in store-gateway: streaming encoding.DecBuf

### ISSUE THREAD

**Issue by 56quarters** [grafana/mimir#3484]
Outlined in [this comment](https://github.com/grafana/mimir/issues/3465#issuecomment-1318850148) are the high level steps to remove mmap. Each of the main bits of functionality in index-header (symbols, posting offsets) requires an `encoding.DecBuf` that operates on a byte slice of the entire index-header. Our first task should be to create a new implementation of `encoding.DecBuf` that uses standard file operations and byte buffers to jump around the contents of a file instead of a byte slice of the entire file.

**Comment by 56quarters** [comment]
@stevesg Are you able to push your experiments with a streaming version of the index-header code somewhere?

**Comment by pracucci** [comment]
> @stevesg Are you able to push your experiments with a streaming version of the index-header code somewhere?

Steve just shared internally, so I'm copy-pasting the reply here (since it's non sensitive):

Half finished prototype of removing mmap from store-gateway is here: https://github.com/grafana/mimir/commit/d8e5035173f13fda7990924a7deda66ec44d5aed

It contains a bunch of stuff:
- A unit test which simulates a slow block device to reliably reproduce slow page faults
- "read_binary_reader.go" - This implements the binary reader but just reading the entire mmap regions into buffers instead of using mmap.
- A fork of of the Decbuf struct which uses a Reader-like interface instead of a byte slice.
- I started work on a new "stream_binary_reader", but never finished it. It uses the stream-based Decbuf, initially just with a memory backing to validate correctness, with the idea it would be swapped out with the file backed stream. It uses a fork of NewSymbols, which I think I got mostly working, albeit with some bug around computing the right offsets.
